File: python/l1_pager/summarizer.py
"""
L1-Pager: Fast extractive summarizer for pointer generation.

Constraint: must complete in <1ms per message. No LLM call.
Produces a 1-sentence description so the model can decide whether
to request the evicted page or ignore it.
"""
from __future__ import annotations

import json
import re
from typing import Any

from langchain_core.messages import AIMessage, BaseMessage, ToolMessage

_SENTENCE_RE = re.compile(r"[.!?]\s")
_MAX_SUMMARY_CHARS = 140


def _first_sentence(text: str) -> str:
    # Collapse newlines so the summary is always a single line
    text = " ".join(text.split())
    m = _SENTENCE_RE.search(text)
    return text[: m.end()].strip() if m else text[:_MAX_SUMMARY_CHARS]


def _summarize_json(text: str) -> str:
    """Try to extract meaningful keys from a JSON blob."""
    try:
        obj = json.loads(text.strip())
        if isinstance(obj, dict):
            keys = list(obj.keys())[:5]
            return f"JSON object with keys: {', '.join(keys)}"
        if isinstance(obj, list):
            return f"JSON array with {len(obj)} items"
    except (json.JSONDecodeError, ValueError):
        pass
    return ""


def _char_count(content: Any) -> int:
    if isinstance(content, str):
        return len(content)
    return len(str(content))


def generate_summary(message: BaseMessage) -> str:
    """
    Generate a ≤140-char, single-line extractive summary for the pointer string.

    Strategy (in order):
    1. If content starts with JSON → extract key names
    2. Extract the first sentence
    3. Fall back to a type+size template

    Invariant: the returned string contains no newline characters.
    """
    content = message.content
    content_str = content if isinstance(content, str) else str(content)
    content_str = content_str.strip()
    char_count = len(content_str)

    raw: str

    # JSON blobs
    if content_str and content_str[0] in ("{", "["):
        summary = _summarize_json(content_str)
        if summary:
            raw = summary[:_MAX_SUMMARY_CHARS]
        else:
            raw = f"{type(message).__name__} ({char_count} chars)"
    elif isinstance(message, ToolMessage) and message.name:
        first = _first_sentence(content_str)
        if len(first) > 15:
            raw = f"{message.name}: {first}"[:_MAX_SUMMARY_CHARS]
        else:
            raw = f"{message.name} output ({char_count} chars)"
    else:
        first = _first_sentence(content_str)
        raw = first[:_MAX_SUMMARY_CHARS] if len(first) > 15 else f"{type(message).__name__} ({char_count} chars)"

    # Invariant: no newlines — a multiline summary would break the pointer regex
    return " ".join(raw.split())

File: typescript/src/summarizer.ts
/**
 * L1-Pager: Fast extractive summarizer (TypeScript)
 * Mirrors python/l1_pager/summarizer.py exactly.
 * Invariant: returned string is single-line, ≤140 chars.
 */

import { BaseMessage, ToolMessage } from "@langchain/core/messages";

const MAX_SUMMARY_CHARS = 140;
const SENTENCE_RE = /[.!?]\s/;

function firstSentence(text: string): string {
  const collapsed = text.replace(/\s+/g, " ").trim();
  const m = SENTENCE_RE.exec(collapsed);
  return m ? collapsed.slice(0, m.index + m[0].length).trim() : collapsed.slice(0, MAX_SUMMARY_CHARS);
}

function summarizeJson(text: string): string {
  try {
    const obj = JSON.parse(text.trim());
    if (obj && typeof obj === "object" && !Array.isArray(obj)) {
      const keys = Object.keys(obj).slice(0, 5).join(", ");
      return `JSON object with keys: ${keys}`;
    }
    if (Array.isArray(obj)) {
      return `JSON array with ${obj.length} items`;
    }
  } catch {
    // not valid JSON
  }
  return "";
}

export function generateSummary(message: BaseMessage): string {
  const rawContent = message.content;
  const contentStr = (typeof rawContent === "string" ? rawContent : JSON.stringify(rawContent)).trim();
  const charCount = contentStr.length;

  let raw: string;

  if (contentStr && (contentStr[0] === "{" || contentStr[0] === "[")) {
    const jsonSummary = summarizeJson(contentStr);
    raw = jsonSummary ? jsonSummary.slice(0, MAX_SUMMARY_CHARS) : `${message._getType()} (${charCount} chars)`;
  } else if (message instanceof ToolMessage && message.name) {
    const first = firstSentence(contentStr);
    raw = first.length > 15
      ? `${message.name}: ${first}`.slice(0, MAX_SUMMARY_CHARS)
      : `${message.name} output (${charCount} chars)`;
  } else {
    const first = firstSentence(contentStr);
    raw = first.length > 15 ? first.slice(0, MAX_SUMMARY_CHARS) : `${message._getType()} (${charCount} chars)`;
  }

  // Invariant: no newlines
  return raw.replace(/\s+/g, " ").trim();
}
